gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/process/traversal/step/map/UnfoldStep.java

```diff
@@ -46,6 +47,8 @@ else if (s instanceof Iterable)
             return ((Iterable) s).iterator();
         else if (s instanceof Map)
             return ((Map) s).entrySet().iterator();
+        else if (s.getClass().isArray())
+            return new ArrayIterator((Object[])s);
```

Comment:
This cast will fail if the incoming array is an array of primitives.  I think you could do something like this which is short and to the point but creates an intermediate, albeit short-lived list.
`return (Iterator<E>)Arrays.asList(s).iterator();`

I was able to reproduce the issue when I nested a primitive array inside of a list.  Calling `unfold` twice looks a little silly but maybe someone would do this if they were dealing with nested collections.

```
gremlin> __.inject([0, 1, 2, [3, 4] as int[]]).unfold().unfold()
==>0
==>1
==>2
[I cannot be cast to [Ljava.lang.Object;
Type ':help' or ':h' for help.
Display stack trace? [yN]n
gremlin> __.inject([0, 1, 2, [3, 4]]).unfold().unfold()
==>0
==>1
==>2
==>3
==>4
```